sunscreen_runtime/src/linked.rs

```diff
@@ -343,10 +294,216 @@ impl<Q: Ring + CryptoHash + ModSwitch<ZqRistretto> + RingModulus<4> + Ord> Linke
 }
+
+impl Sdlp {
+    const TRANSCRIPT_LABEL: &'static [u8] = b"solo-sdlp";
+    /// This function creates a singular SDLP, not linked to any other proof system. This can be
+    /// used when only proving valid encryptions of known values, but _not_ for proving any
+    /// properties of those underlying values.
+    ///
+    /// The [builder methods](`crate::LogProofBuilder`) offer an easier way to construct this proof.
+    pub fn create(prover_knowledge: &SealSdlpProverKnowledge) -> Result<Self> {
+        let mut transcript = Transcript::new(Self::TRANSCRIPT_LABEL);
+        let vk = prover_knowledge.vk();
+        let gen = LogProofGenerators::new(vk.l() as usize);
+        let u = InnerProductVerifierKnowledge::get_u();
+        let proof = prover_knowledge.create_logproof(&mut transcript, &gen.g, &gen.h, &u);
+
+        Ok(Self {
+            proof,
+            vk,
+            g: gen.g,
+            h: gen.h,
+            u,
+        })
+    }
+
+    /// This function verifies a solo SDLP.
+    pub fn verify(&self) -> Result<()> {
+        let mut transcript = Transcript::new(Self::TRANSCRIPT_LABEL);
+
+        self.vk
+            .verify(&self.proof, &mut transcript, &self.g, &self.h, &self.u)?;
+
+        Ok(())
+    }
+}
+
+/// The prover knowledge of an [`Sdlp`].
+pub struct SealSdlpProverKnowledge(pub(crate) SealSdlpProverKnowledgeInternal);
+
+/// The verifier knowledge of an [`Sdlp`].
+#[derive(Debug, Clone)]
+pub struct SealSdlpVerifierKnowledge(pub(crate) SealSdlpVerifierKnowledgeInternal);
+
+pub(crate) enum SealSdlpProverKnowledgeInternal {
+    LP1(LogProofProverKnowledge<ZqSeal128_1024>),
+    LP2(LogProofProverKnowledge<ZqSeal128_2048>),
+    LP3(LogProofProverKnowledge<ZqSeal128_4096>),
+    LP4(LogProofProverKnowledge<ZqSeal128_8192>),
+}
+
+#[derive(Debug, Clone)]
+pub(crate) enum SealSdlpVerifierKnowledgeInternal {
+    LP1(LogProofVerifierKnowledge<ZqSeal128_1024>),
+    LP2(LogProofVerifierKnowledge<ZqSeal128_2048>),
+    LP3(LogProofVerifierKnowledge<ZqSeal128_4096>),
+    LP4(LogProofVerifierKnowledge<ZqSeal128_8192>),
+}
+
+macro_rules! impl_from {
+    ($zq_type:ty, $variant:ident) => {
+        paste! {
+            impl From<LogProofProverKnowledge<$zq_type>> for SealSdlpProverKnowledge {
+                fn from(k: LogProofProverKnowledge<$zq_type>) -> Self {
+                    Self(SealSdlpProverKnowledgeInternal::$variant(k))
+                }
+            }
+            impl From<LogProofVerifierKnowledge<$zq_type>> for SealSdlpVerifierKnowledge {
+                fn from(k: LogProofVerifierKnowledge<$zq_type>) -> Self {
+                    Self(SealSdlpVerifierKnowledgeInternal::$variant(k))
+                }
+            }
+        }
+    };
+}
+
+impl_from!(ZqSeal128_1024, LP1);
+impl_from!(ZqSeal128_2048, LP2);
+impl_from!(ZqSeal128_4096, LP3);
+impl_from!(ZqSeal128_8192, LP4);
+
+macro_rules! seq_zq {
+    ($block:tt) => (
+        seq!(N in 1..=4 {
+            $block
+        })
+    )
+}
```

Comment:
What is this doing?

Reply:
This is making the manual dynamic dispatch wrappers more maintainable. I had to create these
```rust
pub(crate) enum SealSdlpProverKnowledgeInternal {
    LP1(LogProofProverKnowledge<ZqSeal128_1024>),
    LP2(LogProofProverKnowledge<ZqSeal128_2048>),
    LP3(LogProofProverKnowledge<ZqSeal128_4096>),
    LP4(LogProofProverKnowledge<ZqSeal128_8192>),
}

#[derive(Debug, Clone)]
pub(crate) enum SealSdlpVerifierKnowledgeInternal {
    LP1(LogProofVerifierKnowledge<ZqSeal128_1024>),
    LP2(LogProofVerifierKnowledge<ZqSeal128_2048>),
    LP3(LogProofVerifierKnowledge<ZqSeal128_4096>),
    LP4(LogProofVerifierKnowledge<ZqSeal128_8192>),
}
```
to go from dynamic parameter sets to compile time parameter sets. That's a lot of match statements and a lot of boilerplate to maintain if we support more/less parameter sets or if we have to expose more functions.